src/coreclr/src/jit/lsra.cpp

```diff
@@ -2994,291 +2989,584 @@ regNumber LinearScan::tryAllocateFreeReg(Interval* currentInterval, RefPosition*
+    regMaskTP lowestCostSpillSet = RBM_NONE;
+    // Apply the SPILL_COST heuristic and eliminate regs that can't be spilled.
+    if (!found)
+    {
+        // The spill weight for 'refPosition' (the one we're allocating now).
+        float thisSpillWeight = getWeight(refPosition);
+        // The  spill weight for the best candidate we've found so far.
+        float bestSpillWeight = FloatingPointUtils::infinite_float();
+        // True if we found registers with lower spill weight than this refPosition.
+        bool foundLowerSpillWeight = false;
+
+        for (regMaskTP spillCandidates = selector.candidates; spillCandidates != RBM_NONE;)
+        {
+            regMaskTP spillCandidateBit = genFindLowestBit(spillCandidates);
+            spillCandidates &= ~spillCandidateBit;
+            regNumber  spillCandidateRegNum    = genRegNumFromMask(spillCandidateBit);
+            RegRecord* spillCandidateRegRecord = &physRegs[spillCandidateRegNum];
+
+            // Can and should the interval in this register be spilled for this one,
+            // if we don't find a better alternative?
+            if ((getNextIntervalRefLocation(spillCandidateRegNum ARM_ARG(regType)) == currentLocation) &&
+                !spillCandidateRegRecord->assignedInterval->getNextRefPosition()->RegOptional())
+            {
+                continue;
+            }
+            if (!isSpillCandidate(currentInterval, refPosition, spillCandidateRegRecord))
+            {
+                continue;
+            }
+
+            float currentSpillWeight = spillCost[spillCandidateRegNum];
+#ifdef TARGET_ARM
+            if (currentInterval->registerType == TYP_DOUBLE)
+            {
+                currentSpillWeight = max(currentSpillWeight, spillCost[REG_NEXT(spillCandidateRegNum)]);
+            }
+#endif
+            if (currentSpillWeight < bestSpillWeight)
+            {
+                bestSpillWeight    = currentSpillWeight;
+                lowestCostSpillSet = spillCandidateBit;
+            }
+            else if (currentSpillWeight == bestSpillWeight)
+            {
+                lowestCostSpillSet |= spillCandidateBit;
+            }
+        }
+
+        // We won't spill if this refPosition is RegOptional() and we have no candidates
+        // with a lower spill cost.
+        if ((bestSpillWeight >= thisSpillWeight) && refPosition->RegOptional())
+        {
+            currentInterval->assignedReg = nullptr;
+            return REG_NA;
+        }
+    }
+
+    if (!found)
+    {
+        // We must have at least one with the lowest spill cost.
+        assert(lowestCostSpillSet != RBM_NONE);
+        found = selector.applySelection(SPILL_COST, lowestCostSpillSet);
+    }
+
+    // Apply the FAR_NEXT_REF heuristic.
+    if (!found)
+    {
+        LsraLocation farthestLocation = MinLocation;
+        regMaskTP    farthestSet      = RBM_NONE;
+        for (regMaskTP farthestCandidates = selector.candidates; farthestCandidates != RBM_NONE;)
```

Comment:
The objective is to make it easier to apply the heuristics in a different order, so that we don't always select a free register. Also, I had hoped that, with the cached values in arrays, we might be able to leverage the native compiler's auto-vectorization capabilities for at least some of the computations.